Add snapshots pending deletion in cluster state to delete snapshot once index is deleted

tlrx opened the pull request:
_Note: This pull request is another attempt to delete searchable snapshot when the mounted index is deleted. It is extracted from #75565 but this one stores the information of the snapshot to delete in the cluster state as `SnapshotDeletionsInPending` custom objects whereas the older ones tried to store those information in the `RepositoryMetadata`._

In #74977 we introduced a new index setting index.store.snapshot.delete_searchable_snapshot that can be set when mounting a snapshot as an index to inform that the snapshot should be deleted once the searchable snapshot index is deleted.

The previous pull request adds the index setting and the verifications around it. This pull request now adds the logic to detect that a searchable snapshot index with this specific logic is being deleted and triggers the deletion of the backing snapshot.

In order to do this, when a searchable snapshot index is deleted we check if the setting index.store.snapshot.delete_searchable_snapshot is set. If the index to be deleted is the last searchable snapshot index that uses the snapshot then the snapshot informations are added to cluster state in a new `SnapshotDeletionsInPending` custom object. 

Once a snapshot is pending deletion it cannot be cloned, mounted or restored in the cluster.

Snapshots pending deletions are deleted by the SnapshotsService. On cluster state updates the SnapshotsService retrieves the list of snapshots to delete and trigger the deletion by executing an explicit snapshot delete request. Deletions of snapshots are executed per repository, the service tries to prevent conflicting situations before triggering deletions.

There are situations were a snapshot pending deletion could not be deleted, for example if a repository is updated to point to a different location or if a repository is set to read-only. In such cases the snapshot information is kept around in `SnapshotDeletionsInPending` so that the deletion can be retried in the future. A hard limit of 5000 pending snapshot is implemented to avoid cluster state 💥 .

tlrx committed:
Add snapshots pending deletion in cluster state to delete snapshot when searchable snapshot index

henningandersen reviewed (commented):
I did an initial read of this and left some initial mostly minor comments. I will need more time to fully digest this.

tlrx committed:
rename to SnapshotDeletionsPending

tlrx committed:
MAX_PENDING_DELETIONS_SETTING

tlrx committed:
list

tlrx committed:
tostring

tlrx committed:
remove if is empty

tlrx committed:
remove if

tlrx committed:
Merge branch 'master' into delete-searchable-snapshot-with-pending-deletes

tlrx committed:
remove triggered

henningandersen reviewed (commented):
I have done a second round of all the production changes, will await a 3rd round before I look into tests too.

tlrx committed:
format conflicting files

tlrx committed:
Merge branch 'master' into delete-searchable-snapshot-with-pending-deletes

tlrx committed:
nits

tlrx committed:
feedback

tlrx committed:
Merge branch 'master' into delete-searchable-snapshot-with-pending-deletes

tlrx committed:
spotless

tlrx committed:
Merge branch 'master' into delete-searchable-snapshot-with-pending-deletes

tlrx committed:
skip cluster state updates

tlrx committed:
add repo clean up test

tlrx committed:
also clean ups

tlrx committed:
Merge branch 'master' into delete-searchable-snapshot-with-pending-deletes

tlrx committed:
retries + expiration

tlrx committed:
Merge branch 'master' into delete-searchable-snapshot-with-pending-deletes

tlrx committed:
missing uuid

tlrx commented on server/src/main/java/org/elasticsearch/cluster/SnapshotDeletionsPending.java:
```diff
@@ -0,0 +1,270 @@
+    );
+
+    /**
+     * A list of snapshots to delete, sorted by creation time
```
OK, I pushed 77dfe0085f8e5179f53749ac41b87e8f38097b8b

tlrx commented on server/src/main/java/org/elasticsearch/cluster/SnapshotDeletionsPending.java:
```diff
@@ -0,0 +1,270 @@
+
+    @Override
+    public Version getMinimalSupportedVersion() {
+        return Version.CURRENT.minimumCompatibilityVersion();
```
Right, I pushed 77dfe0085f8e5179f53749ac41b87e8f38097b8b

tlrx commented on server/src/main/java/org/elasticsearch/cluster/SnapshotDeletionsPending.java:
```diff
@@ -0,0 +1,270 @@
+        boolean changed = false;
+        final List<Entry> updatedEntries = new ArrayList<>();
+        for (Entry entry : entries) {
+            if (snapshotIds.contains(entry.snapshotId)) {
```
Makes sense, I changed this in 77dfe0085f8e5179f53749ac41b87e8f38097b8b.

tlrx commented on server/src/main/java/org/elasticsearch/cluster/SnapshotDeletionsPending.java:
```diff
@@ -0,0 +1,270 @@
+            prepend = false;
+        }
+        builder.append(']');
+        return builder.toString();
```
I used this to add extra information when debugging, but it's not needed anymore. I changed to what you suggested as part of 77dfe0085f8e5179f53749ac41b87e8f38097b8b.

tlrx commented on server/src/main/java/org/elasticsearch/cluster/SnapshotDeletionsPending.java:
```diff
@@ -0,0 +1,270 @@
+    public static final String TYPE = "snapshot_deletions_pending";
+
+    /**
+     * Setting for the maximum number of snapshots pending deletion allowed in the cluster state.
```
I improved the comment in d15ac48d5d2043171bcbcf3f1fb0c179969cb0bb

tlrx commented on server/src/main/java/org/elasticsearch/cluster/SnapshotDeletionsPending.java:
```diff
@@ -0,0 +1,270 @@
+        private final String repositoryName;
+        private final String repositoryUuid;
+        private final SnapshotId snapshotId;
+        private final long creationTime;
```
Sure, I renamed it to `indexDeletionTime` in 77dfe0085f8e5179f53749ac41b87e8f38097b8b.

tlrx commented on server/src/main/java/org/elasticsearch/cluster/metadata/MetadataDeleteIndexService.java:
```diff
@@ -120,26 +133,115 @@ public ClusterState deleteIndices(ClusterState currentState, Set<Index> indices)
+                indexSettings.get(SEARCHABLE_SNAPSHOTS_SNAPSHOT_UUID_SETTING_KEY)
+            );
+            boolean canDeleteSnapshot = true;
+            for (IndexMetadata other : metadata) {
```
I changed this in ea065e95641f4df416395e6d3279027b7450eaf7 (sorry, another large change) to build the list of index settings that are going to be deleted and the list of active searchable snapshots. This way we can more easily compare the searchable snapshots that are still in use vs. the ones that needs to be deleted.

tlrx commented on server/src/main/java/org/elasticsearch/repositories/RepositoriesService.java:
```diff
@@ -350,6 +350,8 @@ public ClusterState execute(ClusterState currentState) {
                             if (Regex.simpleMatch(request.name(), repositoryMetadata.name())) {
                                 ensureRepositoryNotInUse(currentState, repositoryMetadata.name());
                                 ensureNoSearchableSnapshotsIndicesInUse(currentState, repositoryMetadata);
+                                // TODO should we prevent the deletion of repositories that have snapshots pending deletions or should
+                                // we just log a warning and rely on SnapshotDeletionsPending to catch up if the repo comes back
```
As we discussed few weeks ago I'll follow up with a change that prevent the deletion/unregistration of repositories with pending snapshot deletions (including a special flag to force the deletion).

tlrx commented on server/src/main/java/org/elasticsearch/cluster/metadata/MetadataDeleteIndexService.java:
```diff
@@ -120,26 +133,115 @@ public ClusterState deleteIndices(ClusterState currentState, Set<Index> indices)
+                canDeleteSnapshot = false; // another index is using the same snapshot, do not delete the snapshot
+                break;
+            }
+            if (canDeleteSnapshot) {
```
I added a comment at the method level as part of ea065e95641f4df416395e6d3279027b7450eaf7.

tlrx commented on server/src/main/java/org/elasticsearch/cluster/SnapshotDeletionsPending.java:
```diff
@@ -0,0 +1,270 @@
+     */
+    public static final Setting<Integer> MAX_PENDING_DELETIONS_SETTING = Setting.intSetting(
+        "cluster.snapshot.snapshot_deletions_pending.size",
+        500,
```
I increased this to 5000 as part of 77dfe0085f8e5179f53749ac41b87e8f38097b8b.

tlrx commented on server/src/main/java/org/elasticsearch/cluster/metadata/MetadataDeleteIndexService.java:
```diff
@@ -120,26 +133,115 @@ public ClusterState deleteIndices(ClusterState currentState, Set<Index> indices)
+            for (IndexMetadata other : metadata) {
+                if (indexToDelete.equals(other.getIndex())) {
+                    continue; // do not check against itself
+                }
```
++ reworked in ea065e95641f4df416395e6d3279027b7450eaf7

tlrx commented on server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:
```diff
@@ -559,6 +576,34 @@ private static void ensureSnapshotNameAvailableInRepo(RepositoryData repositoryD
         }
     }
 
+    private static Set<SnapshotId> listOfCloneSources(final ClusterState state) {
```
I renamed the methods in 77dfe0085f8e5179f53749ac41b87e8f38097b8b and used streams + filters to make the method more similar.

tlrx commented on server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:
```diff
@@ -559,6 +576,34 @@ private static void ensureSnapshotNameAvailableInRepo(RepositoryData repositoryD
+    }
+
+    private static Set<SnapshotId> listOfRestoreSources(final ClusterState state) {
+        final Set<SnapshotId> snapshotIds = new HashSet<>();
```
I renamed the methods in 77dfe00 and used streams + filters to make the method more similar.

tlrx commented on server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:
```diff
@@ -1245,6 +1296,258 @@ public void clusterStateProcessed(String source, ClusterState oldState, ClusterS
+     *
+     * @param state the current {@link ClusterState}
+     */
+    private void triggerSnapshotsPendingDeletions(final ClusterState state) {
```
Ok, this suggestion is a very good one but it required a larger change than I expected (see 916689e76573fc1c93c8403da91a808cc838e1ab).

I reworked this and introduced the `pendingDeletionsChanged(ClusterChangedEvent)` and `pendingDeletionsWithConflictsChanged(ClusterChangedEvent)` methods that checks the cluster state update and returns `true` if it is worth to iterate over the pending snapshot deletions. Those methods are based on `Set<SnapshotId>` that are kept and updated locally on the master node. Those sets are used to know if a pending snapshot deletion is waiting for a conflicting situation, for example a restore, to be completed.

tlrx commented on server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:
```diff
@@ -1245,6 +1296,258 @@ public void clusterStateProcessed(String source, ClusterState oldState, ClusterS
+                            shouldRetry = RepositoryData.MISSING_UUID.equals(repositoryUuid) == false;
+
+                        } else if (e instanceof ConcurrentSnapshotExecutionException) {
+                            logger.debug(
```
Yes, we handle concurrent deletes silently but I think we can still get there with concurrent repository clean ups. I changed this code to handle all exceptions the same way (except snapshot missing exception) in cf66c462179b0a1400ecb6b6637ecc6de456ab5a.

tlrx commented on server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:
```diff
@@ -1245,6 +1296,258 @@ public void clusterStateProcessed(String source, ClusterState oldState, ClusterS
+                                snapshotId
+                            );
+                        } else if (e instanceof RepositoryMissingException) {
+                            logger.warn(
```
> It ties in with how we handle repo deletions. If we only allow forcing those when there are pending deletes, we could perhaps log (and respond) the these snapshots at that time.

I think this is a good suggestion. I'd like to do this in a follow up.

> we could log the snapshots when adding to the pending deletions list if the repo is not even there when the index is deleted

I added a warning in `MetadataDeleteIndexService#updateSnapshotDeletionsPending()` for this.

tlrx commented on server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:
```diff
@@ -1245,6 +1296,258 @@ public void clusterStateProcessed(String source, ClusterState oldState, ClusterS
+                continue;
+            }
+
+            // should we add some throttling to not always retry?
```
I implemented a basic retry mechanism in cf66c462179b0a1400ecb6b6637ecc6de456ab5a that retries all exceptions at a constant interval (30s by default) over a given expiration time after which the retries will be stopped and the pending snapshot deletion removed from the cluster state with a warning message.

This expiration delay is only evaluated after the pending deletion has been already triggered and failed. It is still possible for a pending deletion to stay in cluster state for a long time if the deletion is blocked by a missing/read only repository or a conflicting operation.

tlrx commented:
@henningandersen thanks a lot for your feedback. I updated the code to address your feedback but it required a much larger change than I initially expected:
- I implemented some methods to avoid evaluating the cluster state updates too frequently but only when pending deletions should be triggered after a conflicting operation is completed or a new deletion has been added
- I implemented a basic retry mechanism to reexecute failed snapshot deletions using a constant interval (30s) and to remove pending deletions that failed to be executed after 12 hours
- In case a pending deletion cannot be triggered in the first place because of a missing/read only repository then the pending deletion is kept in cluster state unless the max. number (5000) is reached and in this case it is simply removed.

I'd appreciate if you can have another look, thanks

tlrx committed:
Merge branch 'master' into delete-searchable-snapshot-with-pending-deletes

tlrx committed:
Merge branch 'master' into delete-searchable-snapshot-with-pending-deletes

tlrx commented:
I opened #81596 for the snapshot test failure which looks unrelated to me.

tlrx committed:
Merge branch 'master' into delete-searchable-snapshot-with-pending-deletes

tlrx commented:
@henningandersen do you think you'll have the time to have another look? I know it's boring and we iterate a lot on this, but I'd be glad to grab your feedback again :)

henningandersen commented on server/src/test/java/org/elasticsearch/cluster/metadata/MetadataDeleteIndexServiceTests.java:
```diff
@@ -212,6 +221,111 @@ public void testDeleteCurrentWriteIndexForDataStream() {
+        if (deleteSnapshot) {
+            assertThat(updatedPendingDeletions, notNullValue());
+            assertThat(updatedPendingDeletions.isEmpty(), equalTo(false));
+            assertThat(updatedPendingDeletions.contains(new SnapshotId("snap_name", "snap_uuid")), equalTo(true));
```
Can we also validate the repo uuid and repo name are correct?

tlrx replied:
Assertions added in fce436543320a8b2e610280cb8328223e49c3fbd, thanks for spotting this.

henningandersen commented on server/src/test/java/org/elasticsearch/cluster/metadata/MetadataDeleteIndexServiceTests.java:
```diff
@@ -212,6 +221,111 @@ public void testDeleteCurrentWriteIndexForDataStream() {
+        final SnapshotDeletionsPending updatedPendingDeletions = updatedState.custom(SnapshotDeletionsPending.TYPE);
+        if (deleteSnapshot) {
+            assertThat(updatedPendingDeletions, notNullValue());
+            assertThat(updatedPendingDeletions.isEmpty(), equalTo(false));
```
Can we also validate the it has size 1, i.e., only led to one pending deletion?

tlrx replied:
Added in fce436543320a8b2e610280cb8328223e49c3fbd

henningandersen commented on server/src/test/java/org/elasticsearch/cluster/metadata/MetadataDeleteIndexServiceTests.java:
```diff
@@ -212,6 +221,111 @@ public void testDeleteCurrentWriteIndexForDataStream() {
+            pendingDeletions = clusterState.custom(SnapshotDeletionsPending.TYPE, SnapshotDeletionsPending.EMPTY);
+        }
+
+        assertThat(pendingDeletions.isEmpty(), equalTo(false));
```
Can we also validate the it has size 1, i.e., only led to one pending deletion?

tlrx replied:
Also added in fce436543320a8b2e610280cb8328223e49c3fbd

henningandersen commented on x-pack/plugin/searchable-snapshots/src/internalClusterTest/java/org/elasticsearch/xpack/searchablesnapshots/SearchableSnapshotsPendingDeletionsIntegTests.java:
```diff
@@ -0,0 +1,575 @@
+        });
+    }
+
+    public void testSnapshotPendingDeletionCannotBeRestored() throws Exception {
```
Do we need a similar test for clone?

tlrx replied:
Sure - I added such a test in df1ec28a06bec4a77d06554cf274348ca9a724fd where I also moved the `ensureSnapshotNotDeletedOrPendingDeletion` into the `SnapshotUtils` class (so that restore and snapshot use the same utility method and have a more coherent error message)

henningandersen commented on x-pack/plugin/searchable-snapshots/src/internalClusterTest/java/org/elasticsearch/xpack/searchablesnapshots/SearchableSnapshotsPendingDeletionsIntegTests.java:
```diff
@@ -0,0 +1,575 @@
+        mountIndexThenExecute((repository, snapshot, index) -> {
+            try {
+                final String masterNode = internalCluster().getMasterName();
+                blockMasterOnShardLevelSnapshotFile(repository, getRepositoryData(repository).resolveIndexId(index).getId());
```
I think the naming of some of the underlying methods is confusing: `setBlockAndOnWriteShardLevelSnapFiles`, just double checking that the effect we are after is to block on reads here?

tlrx replied:
I opened #82530 to rename the method.

tlrx replied:
(and yes, we only want to block reads here)

henningandersen commented on x-pack/plugin/searchable-snapshots/src/internalClusterTest/java/org/elasticsearch/xpack/searchablesnapshots/SearchableSnapshotsPendingDeletionsIntegTests.java:
```diff
@@ -0,0 +1,575 @@
+        }
+        mountLatch.await();
+
+        if (randomBoolean()) {
+            // we can't delete a repository that is used by searchable snapshot indices but we can update it
+            assertAcked(
+                clusterAdmin().preparePutRepository(repository)
+                    .setType(FsRepository.TYPE)
+                    .setVerify(randomBoolean())
+                    .setSettings(randomBoolean() ? repositorySettings : repositorySettings.put("dummy", randomInt()))
+            );
+        }
```
I am not exactly sure what this bit validates? Perhaps add a comment?

tlrx replied:
I added a comment in 2d8b32a649e777a273253716bbaccc8303f462fe:
>            // Force the re-registration of the repository with randomized value for the "verify" flag;
>             // it helps to test the case where the repository UUID is unknown at the time the snapshot
>             // is marked to be deleted.

henningandersen commented on x-pack/plugin/searchable-snapshots/src/internalClusterTest/java/org/elasticsearch/xpack/searchablesnapshots/SearchableSnapshotsPendingDeletionsIntegTests.java:
```diff
@@ -0,0 +1,575 @@
+        createRepository(repository, FsRepository.TYPE, repositorySettings);
+
+        final List<String> indices = new ArrayList<>();
+        for (int i = 0; i < randomIntBetween(1, 10); i++) {
```
This will bias towards smaller lists, since the `randomIntBetween` is evaluated for every loop iteration. I think you intended to make this constant instead? Otherwise it deserves a comment.

henningandersen commented on x-pack/plugin/searchable-snapshots/src/internalClusterTest/java/org/elasticsearch/xpack/searchablesnapshots/SearchableSnapshotsPendingDeletionsIntegTests.java:
```diff
@@ -0,0 +1,575 @@
+                do {
+                    List<String> deletions = randomSubsetOf(Math.max(1, mountedIndices.size() - 1), mountedIndices);
+                    assertAcked(client().admin().indices().prepareDelete(deletions.toArray(String[]::new)));
+                    deletions.forEach(mountedIndices::remove);
```
Is this not the same as `mountedIndices.removeAll(deletions)`?

tlrx replied:
AbstractSet#removeAll's javadoc indicates that the method behaves differently depending of the sizes of the sets that are involved, but in this case `deletions` should be smaller that `mountedIndices` so `removeAll` should operate similarly to `deletions.forEach()`.

henningandersen commented on x-pack/plugin/searchable-snapshots/src/internalClusterTest/java/org/elasticsearch/xpack/searchablesnapshots/SearchableSnapshotsPendingDeletionsIntegTests.java:
```diff
@@ -0,0 +1,575 @@
+                startThreads.await();
+                final Set<String> mountedIndices = new HashSet<>(mounts.keySet());
+                do {
+                    List<String> deletions = randomSubsetOf(Math.max(1, mountedIndices.size() - 1), mountedIndices);
```
Could we randomly pick the size between 1 and `mountedIndices.size()` instead? That would sometimes do them all in one go and sometimes in multiple rounds rather than always two rounds when `mountedIndices.size() > 1`?

tlrx replied:
Yes, that makes more sense => ad62175111bfbf16472c604c66e44520ce90fb9a

henningandersen commented on x-pack/plugin/searchable-snapshots/src/internalClusterTest/java/org/elasticsearch/xpack/searchablesnapshots/SearchableSnapshotsPendingDeletionsIntegTests.java:
```diff
@@ -0,0 +1,575 @@
+        });
+    }
+
+    public void testSearchableSnapshotIsDeletedAfterExpiration() throws Exception {
```
Can we squeeze in "pending deletion" in the name, since it is the pending deletion that is deleted after expiration?

tlrx replied:
Sure, I renamed to `testSnapshotDeletionsPendingIsRemovedFromClusterStateAfterExpiration` in dd1ced9a409ad9490c9e435b674fa3608eebc58a

henningandersen reviewed (commented):
Sorry for the delay here @tanguy. This overall looks good. I left a number of comments, but mostly smaller things.

Also, I want to double check that this is not enabled in ILM yet?

tlrx committed:
Merge branch 'master' into delete-searchable-snapshot-with-pending-deletes

tlrx committed:
5000

tlrx committed:
order of fields

tlrx committed:
Optional.or()

tlrx committed:
remove //TODO

tlrx committed:
remove curly

tlrx committed:
currents

tlrx committed:
no timeout

tlrx committed:
pass null

tlrx committed:
test assertions

tlrx committed:
clone test

tlrx committed:
comment

tlrx committed:
randomInt()

tlrx committed:
randomSubsetOf()

tlrx committed:
add missing setting

tlrx committed:
add SnapshotDeletionsPendingExecutor

tlrx committed:
Merge branch 'master' into delete-searchable-snapshot-with-pending-deletes

tlrx committed:
unbatched

tlrx committed:
same error message

tlrx commented:
@elasticmachine run elasticsearch-ci/part-1

tlrx committed:
Merge branch 'master' into delete-searchable-snapshot-with-pending-deletes

tlrx committed:
fixes

tlrx commented on server/src/main/java/org/elasticsearch/cluster/SnapshotDeletionsPending.java:
```diff
@@ -0,0 +1,281 @@
+ * delete requests to effectively delete the snapshot from the repository. It is possible that the deletion of a snapshot failed for various
+ * reason (ex: conflicting snapshot operation, repository removed etc). In such cases the snapshot pending deletion is kept in the cluster
+ * state and the deletion will be retried on the next cluster state update. To avoid too many snapshots pending deletion stored in cluster
+ * state the number is limited to 500 and configurable through the {@link #MAX_PENDING_DELETIONS_SETTING} setting.
```
I updated the PR description and the doc in 3bd59a3c3f67e928e8240ed78736b50efc7a9a41

tlrx commented on server/src/main/java/org/elasticsearch/cluster/SnapshotDeletionsPending.java:
```diff
@@ -0,0 +1,281 @@
+            this.repositoryName = in.readString();
+            this.repositoryUuid = in.readString();
+            this.indexDeletionTime = in.readVLong();
+            this.snapshotId = new SnapshotId(in);
```
I pushed 4b9080a2f4943f572e64ac9f5e073cc6398ff8e7

tlrx commented on server/src/main/java/org/elasticsearch/cluster/metadata/MetadataDeleteIndexService.java:
```diff
@@ -134,27 +150,137 @@ public ClusterState deleteIndices(ClusterState currentState, Set<Index> indices)
+        final int maxPendingDeletions = SnapshotDeletionsPending.MAX_PENDING_DELETIONS_SETTING.get(settings);
+        final SnapshotDeletionsPending.Builder builder = new SnapshotDeletionsPending.Builder(
+            pendingDeletions,
+            evicted -> logger.warn(
```
> But would we want to fail deletion of indices perhaps if we do too many in one go too fast somehow?
I think we should, yes.

> Perhaps we should fail the deletion if one of the snapshots we cannot delete is one of the ones we pass in here? 
Same.

> I think we can defer this to follow-ups if you prefer.

Yes, I'd like to make this change separately; this PR is already huge.

tlrx commented on server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:
```diff
@@ -1261,6 +1318,413 @@ public void clusterStateProcessed(String source, ClusterState oldState, ClusterS
+                        .filter(repo -> Objects.equals(repo.uuid(), RepositoryData.MISSING_UUID))
+                        .filter(repo -> Objects.equals(repo.name(), repositoryName))
+                        .findFirst();
+                }
```
Sure, I pushed 44552c58aa95590cbcbf5e256fb6ebbefe2109ca

tlrx commented on server/src/main/java/org/elasticsearch/cluster/metadata/MetadataDeleteIndexService.java:
```diff
@@ -134,27 +150,137 @@ public ClusterState deleteIndices(ClusterState currentState, Set<Index> indices)
+            builder.add(entry.getValue().v1(), entry.getValue().v2(), entry.getKey(), timestamp);
+        }
+        for (Map.Entry<SnapshotId, Tuple<String, String>> entry : snapshotsWithoutRepository.entrySet()) {
+            // TODO also log that it will stay as pending for a given time/attempts and then be removed?
```
I removed the TODO in bd863e9f86ff06e442d2cc6a451a25a7df07b8e9

tlrx commented on server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:
```diff
@@ -964,16 +1018,19 @@ public void applyClusterState(ClusterChangedEvent event) {
+                    for (Snapshot snapshot : Set.copyOf(snapshotCompletionListeners.keySet())) {
+                        if (endingSnapshots.add(snapshot)) {
+                            failSnapshotCompletionListeners(snapshot, new SnapshotException(snapshot, "no longer master"));
+                        }
```
This is a left over from some debugging done for #82361. I reverted the change in 8bb56d3a40fb97bdf7a3a36f915db561c75fb3a7

tlrx commented on server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:
```diff
@@ -1261,6 +1318,413 @@ public void clusterStateProcessed(String source, ClusterState oldState, ClusterS
+    private boolean pendingDeletionsWithConflictsChanged(ClusterChangedEvent event) {
+        if (pendingDeletionsWithConflictingRestores.isEmpty() == false) {
+            RestoreInProgress previous = event.previousState().custom(RestoreInProgress.TYPE, RestoreInProgress.EMPTY);
+            RestoreInProgress currents = event.state().custom(RestoreInProgress.TYPE, RestoreInProgress.EMPTY);
```
Sure, I renamed in e541a4cc8fdbbc7c25aae0386ab38ed573e09ff7

tlrx committed:
remove obsolete doc

tlrx commented on server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:
```diff
@@ -1261,6 +1318,413 @@ public void clusterStateProcessed(String source, ClusterState oldState, ClusterS
+     *
+     * @param event the current {@link ClusterChangedEvent}
+     */
+    private void triggerSnapshotsPendingDeletions(final ClusterChangedEvent event) {
```
Right, I should have added the unit tests in the first place.

I extracted the logic of deleting the `SnapshotDeletionsPending` into a dedicated `SnapshotDeletionsPendingExecutor` so that the logic is more contained and more easily testable using unit tests without having to create a complete `SnapshotService`. 

Unit tests have been added in `SnapshotDeletionsPendingExecutorTests` where we check that deletion are correctly triggered after cluster state updates and that conflict are indeed resolved.

tlrx commented on server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:
```diff
@@ -1261,6 +1318,413 @@ public void clusterStateProcessed(String source, ClusterState oldState, ClusterS
+                    deleteSnapshotsByUuid(
+                        repositoryName,
+                        new String[] { snapshotId.getUUID() },
+                        DeleteSnapshotRequest.DEFAULT_MASTER_NODE_TIMEOUT,
```
I agree and removed the timeout in 2188a703442dc0769e5b9141bd5c41b2058e7e5d.

tlrx commented on server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java:
```diff
@@ -2523,7 +2524,8 @@ private ClusterState updateRepositoryGenerationsIfNecessary(ClusterState state,
         updatedDeletionsInProgress = changedDeletions ? SnapshotDeletionsInProgress.of(deletionEntries) : null;
-        return SnapshotsService.updateWithSnapshots(state, updatedSnapshotsInProgress, updatedDeletionsInProgress);
+        final SnapshotDeletionsPending pendingDeletions = state.custom(SnapshotDeletionsPending.TYPE);
+        return SnapshotsService.updateWithSnapshots(state, updatedSnapshotsInProgress, updatedDeletionsInProgress, pendingDeletions);
```
Sure, I pushed 69608bf8e1610ec246a33df7deca48ad5920d41e

tlrx committed:
rename method

tlrx commented:
Thanks @henningandersen for your last review, and sorry it took me so long to update this pull request.

I updated the code with your feedback; I extracted the logic added by this change into a dedicated `SnapshotDeletionsPendingExecutor` class as I finally thought it makes things separated and more easy to maintain and to test. The overall logic did not change since your last review.

I'd be very happy if you can have another look to this, thanks

tlrx commented on server/src/main/java/org/elasticsearch/snapshots/SnapshotDeletionsPendingExecutor.java:
```diff
@@ -0,0 +1,451 @@
+import static org.elasticsearch.snapshots.SnapshotUtils.restoreSources;
+import static org.elasticsearch.snapshots.SnapshotsService.findRepositoryForPendingDeletion;
+
+public class SnapshotDeletionsPendingExecutor {
```
This is an extraction of the previous logic from `SnapshotService.triggerPendingSnapshotDeletion()`

tlrx commented on server/src/main/java/org/elasticsearch/snapshots/SnapshotDeletionsPendingExecutor.java:
```diff
@@ -0,0 +1,451 @@
+     * Map of pending snapshots deletions whose deletion is conflicting with on-going restores/clones/repository clean up or repository
+     * missing or read-only. Those sets are used to identify the cluster state updates to process in case they resolve some conflict.
+     */
+    private final Map<SnapshotId, ConflictType> conflicting = new HashMap<>();
```
This replaces the different sets for the various type of conflicts that were introduced in this PR.

tlrx committed:
Update docs/changelog/79156.yaml

elasticsearchmachine commented:
Hi @tlrx, I've created a changelog YAML for you.

tlrx committed:
Merge branch 'master' into delete-searchable-snapshot-with-pending-deletes

tlrx committed:
batch cluster state updates that removes pending deletions

tlrx committed:
nits

tlrx committed:
Merge branch 'master' into delete-searchable-snapshot-with-pending-deletes

tlrx committed:
reset (wip)

tlrx committed:
update branch

DaveCTurner commented:
I wonder if this'd be simpler if we turn it around and instead introduce a call to `MetadataDeleteIndexService#deleteIndices(ClusterState, ...)` in the cluster state update that commits the snapshot deletion (i.e. the one submitted in `BlobStoreRepository#writeIndexGen`) so that the index deletion happens in the same cluster state update.

If we did that it'd mostly work ok except for the period where we've written an updated `index-N` blob but not updated the repository generation in the cluster state. If there's a master failover during this period then the new master can at least tell it has to do some fixup (the repository generations in the cluster state are not equal) and I think we can include in this fixup process a check for indices-to-be-deleted so we don't miss any.

elasticsearchmachine commented:
Pinging @elastic/es-distributed-obsolete (Team:Distributed (Obsolete))

elasticsearchmachine commented:
Pinging @elastic/es-distributed-coordination (Team:Distributed Coordination)
